alcs-frontend/src/app/features/application/documents/document-upload-dialog/document-upload-dialog.component.ts

```diff
@@ -0,0 +1,166 @@
+    visibleToPublic: this.visibleToPublic,
+  });
+
+  pendingFile: File | undefined;
+  existingFile: { name: string; size: number } | undefined;
+
+  constructor(
+    @Inject(MAT_DIALOG_DATA) public data: { fileId: string; existingDocument?: ApplicationDocumentDto },
+    protected dialog: MatDialogRef<any>,
+    private applicationDocumentService: ApplicationDocumentService
+  ) {}
+
+  ngOnInit(): void {
+    this.loadDocumentTypes();
+
+    if (this.data.existingDocument) {
+      const document = this.data.existingDocument;
+      this.title = 'Edit';
+      this.form.patchValue({
+        name: document.fileName,
+        type: document.type?.code,
+        source: document.source,
+        visibleToApplicant: document.visibilityFlags.includes('A'),
+        visibleToCommissioner: document.visibilityFlags.includes('C'),
+        visibleToPublic: document.visibilityFlags.includes('P'),
+      });
+      this.documentTypeAhead = document.type!.code;
+      this.existingFile = {
+        name: document.fileName,
+        size: 0,
+      };
+    }
+  }
+
+  async onSubmit() {
+    const visibilityFlags: ('A' | 'C' | 'G' | 'P')[] = [];
+    if (this.visibleToApplicant.getRawValue()) {
+      visibilityFlags.push('A');
+    }
+
+    if (this.visibleToCommissioner.getRawValue()) {
+      visibilityFlags.push('C');
+      visibilityFlags.push('G');
+    }
+
+    if (this.visibleToPublic.getRawValue()) {
+      visibilityFlags.push('P');
+    }
+
+    const dto = {
+      file: this.pendingFile,
+      fileName: this.name.getRawValue()!,
+      source: this.source.getRawValue() as DOCUMENT_SOURCE,
+      typeCode: this.type.getRawValue() as DOCUMENT_TYPE,
+      visibilityFlags,
+    };
+    if (this.data.existingDocument) {
+      await this.applicationDocumentService.update(this.data.existingDocument.uuid, dto);
+    } else if (dto.file !== undefined) {
+      // @ts-ignore File is checked above to not be undefined, typescript still thinks it might be undefined
+      await this.applicationDocumentService.upload(this.data.fileId, dto);
+    }
+    this.dialog.close(true);
+  }
+
+  private async loadDocumentTypes() {
+    const docTypes = await this.applicationDocumentService.fetchTypes();
+    docTypes.sort((a, b) => (a.label > b.label ? 1 : -1));
+    this.documentTypes = docTypes;
+  }
+
+  ngOnDestroy(): void {
+    this.$destroy.next();
+    this.$destroy.complete();
+  }
+
+  filterDocumentTypes(term: string, item: ApplicationDocumentTypeDto) {
+    const termLower = term.toLocaleLowerCase();
+    return (
+      item.label.toLocaleLowerCase().indexOf(termLower) > -1 ||
+      item.oatsCode.toLocaleLowerCase().indexOf(termLower) > -1
+    );
+  }
+
+  onDocTypeSelected($event?: ApplicationDocumentTypeDto) {
+    if ($event) {
+      this.type.setValue($event.code);
+    } else {
+      this.type.setValue(undefined);
+    }
+  }
+
+  uploadFile($event: Event) {
+    // @ts-ignore
+    const file: File | undefined = $event.target?.files[0];
```

Comment:
just a nitpick const file?: File. Also what is the reason you need ts-ignore here?